### PR TITLE
mes-2989 Clean up calls to .unsubscribe from observable subscriptions

### PR DESCRIPTION
**Pull request by ataylorBJSS**
## Description and relevant Jira numbers
Refactor subscribe and unsubscribe on observables to clean up on ionViewDidLeave and to subscribe on ionViewWillEnter. If ngOnInit is fired on each page request, the subscribe has been left as is.
[https://jira.i-env.net/browse/MES-2989](https://jira.i-env.net/browse/MES-2989)

## Pull Request checklist:

- [x] [WIP] tag removed from PR title
- [x] PR has an understandable description

## Git feature branch checklist:

- [x] branch name comply with our branching strategy
- [x] git branch contains relevant JIRA ticket number
- [x] branch rebased against the latest develop
- [x] commits are squashed

## Sign off process checklist:

- [ ] Code has been tested manually
- [x] Tests are passing
- [x] PR link added to JIRA
- [x] Reviewers selected in Github
- [ ] Any new reusable component/widget added to component library on Confluence

- [ ] Screenshot(s) captured on the physical device (if applicable)
- [ ] Tested by QA
- [ ] PO's approval

### CODE REVIEW

File: src/pages/test-report/components/manoeuvre-competency/manoeuvre-competency.ts [hunk 1]
```diff
@@ -34,7 +34,7 @@ interface ManoeuvreCompetencyComponentState {
   selector: 'manoeuvre-competency',
   templateUrl: 'manoeuvre-competency.html',
 })
-export class ManoeuvreCompetencyComponent implements OnInit, OnDestroy {
+export class ManoeuvreCompetencyComponent implements OnInit {
```
Comment (johnmorleybjss) [inline src/pages/test-report/components/manoeuvre-competency/manoeuvre-competency.ts]: See you've removed declaration that this component implements `OnDestroy`, but we still use that method. What's the reason for this?

File: src/pages/test-report/components/vehicle-check/vehicle-check.ts [hunk 2]
```diff
@@ -21,12 +21,13 @@ import { isSeriousMode, isDangerousMode, isRemoveFaultMode } from '../../test-re
...
 import { merge } from 'rxjs/observable/merge';
 import { isEmpty } from 'lodash';
+import { Observable } from 'rxjs/Observable';
 
 @Component({
   selector: 'vehicle-check',
   templateUrl: 'vehicle-check.html',
 })
-export class VehicleCheckComponent implements OnInit {
+export class VehicleCheckComponent implements OnInit, OnDestroy {
```
Comment (johnmorleybjss) [inline src/pages/test-report/components/vehicle-check/vehicle-check.ts]: Then declaration of `OnDestroy` has been added here - inconsistent with the above comment.

File: src/pages/view-test-result/view-test-result.ts [hunk 3]
```diff
@@ -56,6 +57,7 @@ export class ViewTestResultPage extends BasePageComponent implements OnInit, OnD
   loadingSpinner: Loading;
   subscription: Subscription;
   showErrorMessage: boolean = false;
+  testResult$ = new Observable<object | StandardCarTestCATBSchema>();
```
Comment (johnmorleybjss) [inline src/pages/view-test-result/view-test-result.ts]: Not sure why this is needed, can you explain?

**Comment by johnmorleybjss** [comment]
As per our conversation, if you find out any more about the differences in behaviour you've encountered (when it's okay to subscribe in `ngOnInit` vs in `ionViewWillEnter`), then please do share with the rest of the team, it'll be good to know.

File: src/pages/test-results-search/test-results-search.ts [hunk 4]
```diff
@@ -34,6 +35,7 @@ export class TestResultsSearchPage extends BasePageComponent {
   hasSearched: boolean = false;
   showSearchSpinner: boolean = false;
   showAdvancedSearchSpinner: boolean = false;
+  subscription: Subscription;
```
Comment (Cavalier07) [inline src/pages/test-results-search/test-results-search.ts]: We can do multiple types searches , without leaving the page. 

The unsubscribe will only work on the last search you did , leaving the others orphaned. Should subscription be an array we push all the searches to and loop through and unsubscribe at the end.

File: src/pages/test-report/components/manoeuvre-competency/manoeuvre-competency.ts (hunk 1 above)
Comment (ataylorBJSS) [reply src/pages/test-report/components/manoeuvre-competency/manoeuvre-competency.ts]: This has just been omitted by mistake. Added back in on last commits

File: src/pages/test-report/components/vehicle-check/vehicle-check.ts (hunk 2 above)
Comment (ataylorBJSS) [reply src/pages/test-report/components/vehicle-check/vehicle-check.ts]: Update the above. Now consistent with this class.

File: src/pages/view-test-result/view-test-result.ts (hunk 3 above)
Comment (ataylorBJSS) [reply src/pages/view-test-result/view-test-result.ts]: It's no longer needed. Removed in latest commit

**Comment by ataylorBJSS** [comment]
## Here's an explanation to the approach taken.

As per [Ionic lifecycle documentation](https://beta.ionicframework.com/docs/building/migration#lifecycle-events) it is recommended to utilise Angular's lifecycle hooks except in certain cases. Therefore I have only used ionic lifecycle hooks were required. 

### Unsubscribing from views.
ionic consistently calls ionViewDidLeave() whenever a page/modal is closed or navigated away from. All unsubscribe methods should be located here.

### Page Components.
Pages pushed onto the navigation stack or navigating back to root (popToRoot())
 - Will always call ngOnInit

Navigating back to pages on the stack
- Should not call ngOnInit() but will call ionViewWillEnter()

### Modals.
Modals will call ngOnit each time they are requested - therefore subscriptions to observables have been left inside this method.

### Page Child Components.
Components within a page component have no concept of the ionic life-cycle as this event is not passed down to them when fired on the page component. Therefore onInit and onDestroy methods are still used for these components.

**Comment by johnmorleybjss** [comment]
Lovely write up, thanks @ataylorBJSS!

File: src/pages/test-results-search/test-results-search.ts (hunk 4 above)
Comment (ataylorBJSS) [reply src/pages/test-results-search/test-results-search.ts]: Thanks Tom,

I've updated the search functions on this to unsubscribe from any previous subscriptions when they are called as well as unsubscribe from the last subscription when leaving the page. Initially setting the subscription to Subscription.Empty means there is no need to check if we have already subscribed to something.

**Pull request merged**